File: apps/gateway/src/openfang/sse.ts
/**
 * SSE consumption for the OpenFang agent stream (BERR-20).
 *
 * `POST /api/agents/{agentId}/message/stream` returns `text/event-stream`. Each
 * frame carries an explicit `event` name and a JSON `data` payload; the pinned
 * stream emits no SSE `id` or retry hints, and non-projectable upstream events
 * arrive as SSE comments (which we drop). This is the compatibility oracle for
 * the Go adapter (`server/internal/openfang/sse.go`), which is the product
 * implementation; both follow the consumption contract:
 *   - projects the five known events into a typed union;
 *   - preserves unknown *named* events as `raw` so nothing is silently lost;
 *   - drops comment/keep-alive frames;
 *   - treats `done` as the end of one model turn, not of the run: the pinned
 *     upstream emits one per turn on the same connection and starts the next
 *     turn whenever the agent called a tool, so the body is read to its end
 *     and nothing is cancelled at `done`;
 *   - treats malformed JSON, a read error, or EOF before any `done` as an
 *     interruption (`STREAM_INTERRUPTED`) so the caller can persist partial
 *     events and mark the Berry run `interrupted` — never re-dispatching, since
 *     the route has no resume/replay contract. EOF after at least one `done`
 *     is the clean end of the run.
 */

import { z } from "zod";
import { OpenFangError } from "~/openfang/errors";

export type OpenFangStreamEvent =
  | { type: "chunk"; content: string }
  | { type: "tool_use"; tool: string }
  | { type: "tool_result"; tool: string; input: Record<string, unknown> }
  | { type: "phase"; phase: string; detail: string | null }
  | { type: "done"; usage: { input_tokens: number; output_tokens: number } }
  | { type: "raw"; event: string; data: unknown };

const chunkDataSchema = z.object({ content: z.string() });
const toolUseDataSchema = z.object({ tool: z.string() });
const toolResultDataSchema = z.object({
  tool: z.string(),
  input: z.record(z.unknown()).default({}),
});
const phaseDataSchema = z.object({
  phase: z.string(),
  detail: z.string().nullable().default(null),
});
const doneDataSchema = z.object({
  usage: z.object({ input_tokens: z.number(), output_tokens: z.number() }),
});

interface RawFrame {
  event: string;
  data: string;
}

/**
 * Consume an OpenFang SSE body, yielding typed events in arrival order until
 * the body ends. Returns normally when the body ends after at least one `done`
 * (each `done` closes a turn; EOF follows the last one); throws `OpenFangError`
 * (`STREAM_INTERRUPTED`) if the stream ends before any `done`, errors, or
 * contains malformed JSON.
 */
export async function* parseOpenFangStream(
  body: ReadableStream<Uint8Array>,
): AsyncGenerator<OpenFangStreamEvent> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  // Counts `done` events. It decides what the end of the body means: after at
  // least one turn it is the run finishing, before any it is the connection
  // dropping mid-work.
  let turns = 0;

  try {
    for (;;) {
      let result: Awaited<ReturnType<typeof reader.read>>;
      try {
        result = await reader.read();
      } catch (cause) {
        throw new OpenFangError("STREAM_INTERRUPTED", "OpenFang stream read failed", { cause });
      }

      if (result.done) {
        buffer += decoder.decode();
        const trailing = buffer.trim();
        if (trailing.length > 0) {
          const event = projectFrame(parseFrame(buffer));
          if (event) {
            if (event.type === "done") turns++;
            yield event;
          }
        }
        break;
      }

      buffer += decoder.decode(result.value, { stream: true });
      const parts = buffer.split(/\r?\n\r?\n/);
      buffer = parts.pop() ?? "";
      for (const block of parts) {
        const event = projectFrame(parseFrame(block));
        if (!event) continue;
        // `done` closes a turn; the next turn arrives on this same connection,
        // so keep reading. Hanging up here left the agent working into a socket
        // nobody read, and its final report never reached Berry.
        if (event.type === "done") turns++;
        yield event;
      }
    }
  } finally {
    reader.releaseLock();
  }

  if (turns === 0) {
    throw new OpenFangError("STREAM_INTERRUPTED", "OpenFang stream ended before any done event");
  }
}

/**
 * Extract the `event` name and joined `data` from one SSE block. Per the SSE
 * spec, multiple `data:` lines are concatenated with `\n`; `:` comment lines and
 * other fields are ignored. Returns `undefined` for comment/keep-alive frames.
 */
export function parseFrame(block: string): RawFrame | undefined {
  let event: string | undefined;
  const dataLines: string[] = [];
  for (const line of block.split(/\r?\n/)) {
    if (line.length === 0 || line.startsWith(":")) continue;
    const separator = line.indexOf(":");
    const field = separator === -1 ? line : line.slice(0, separator);
    // A leading space after the colon is part of the SSE framing, not the value.
    let value = separator === -1 ? "" : line.slice(separator + 1);
    if (value.startsWith(" ")) value = value.slice(1);
    if (field === "event") event = value;
    else if (field === "data") dataLines.push(value);
  }
  if (event === undefined || dataLines.length === 0) return undefined;
  return { event, data: dataLines.join("\n") };
}

/** Parse a raw frame's JSON and project it into a typed event. */
function projectFrame(frame: RawFrame | undefined): OpenFangStreamEvent | undefined {
  if (!frame) return undefined;

  let data: unknown;
  try {
    data = JSON.parse(frame.data);
  } catch (cause) {
    throw new OpenFangError(
      "STREAM_INTERRUPTED",
      `OpenFang stream ${frame.event} event contained malformed JSON`,
      { cause },
    );
  }

  switch (frame.event) {
    case "chunk": {
      const parsed = chunkDataSchema.safeParse(data);
      return parsed.success
        ? { type: "chunk", content: parsed.data.content }
        : { type: "raw", event: frame.event, data };
    }
    case "tool_use": {
      const parsed = toolUseDataSchema.safeParse(data);
      return parsed.success
        ? { type: "tool_use", tool: parsed.data.tool }
        : { type: "raw", event: frame.event, data };
    }
    case "tool_result": {
      const parsed = toolResultDataSchema.safeParse(data);
      return parsed.success
        ? { type: "tool_result", tool: parsed.data.tool, input: parsed.data.input }
        : { type: "raw", event: frame.event, data };
    }
    case "phase": {
      const parsed = phaseDataSchema.safeParse(data);
      return parsed.success
        ? { type: "phase", phase: parsed.data.phase, detail: parsed.data.detail }
        : { type: "raw", event: frame.event, data };
    }
    case "done": {
      const parsed = doneDataSchema.safeParse(data);
      return parsed.success
        ? { type: "done", usage: parsed.data.usage }
        : { type: "raw", event: frame.event, data };
    }
    default:
      return { type: "raw", event: frame.event, data };
  }
}

File: server/internal/openfang/sse.go
package openfang

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
	"sync"
)

// StreamEventType is the bounded pinned upstream SSE event vocabulary.
type StreamEventType string

const (
	EventChunk      StreamEventType = "chunk"
	EventToolUse    StreamEventType = "tool_use"
	EventToolResult StreamEventType = "tool_result"
	EventPhase      StreamEventType = "phase"
	EventDone       StreamEventType = "done"
	EventUnknown    StreamEventType = "unknown"
)

// Usage is the token count one done event reports for the turn it closes.
// The pinned upstream emits done at the end of every model turn, not once per
// run, so a run's total is the sum over its done events.
type Usage struct {
	InputTokens  int64 `json:"input_tokens"`
	OutputTokens int64 `json:"output_tokens"`
}

// StreamEvent contains one validated upstream event. Raw is bounded and is
// populated only for unknown named events; callers must not expose it directly.
type StreamEvent struct {
	Type      StreamEventType
	Content   string
	Tool      string
	Input     json.RawMessage
	Phase     string
	Detail    *string
	Usage     Usage
	EventName string
	Raw       json.RawMessage
}

// StreamErrorKind classifies stream failures without exposing raw payloads.
type StreamErrorKind string

const (
	StreamInterrupted StreamErrorKind = "interrupted"
	StreamMalformed   StreamErrorKind = "malformed"
	StreamLimit       StreamErrorKind = "limit_exceeded"
	StreamTimeout     StreamErrorKind = "timeout"
)

// ErrStreamInterrupted is matched for every non-successful stream ending.
var ErrStreamInterrupted = errors.New("runtime stream interrupted")

// StreamError carries only a stable classification and upstream request ID.
type StreamError struct {
	Kind      StreamErrorKind
	RequestID string
}

func (err *StreamError) Error() string {
	return fmt.Sprintf("runtime stream failed (%s)", err.Kind)
}

func (err *StreamError) Unwrap() error {
	return ErrStreamInterrupted
}

// MessageStream reads one bounded SSE event at a time.
type MessageStream struct {
	ctx        context.Context
	cancel     context.CancelFunc
	body       io.ReadCloser
	scanner    *bufio.Scanner
	requestID  string
	maxTotal   int64
	maxEvent   int
	total      int64
	eventName  string
	dataLines  []string
	frameBytes int
	eof        bool
	// turns counts done events. It decides what the end of the body means:
	// after at least one turn it is the run finishing, before any it is the
	// connection dropping mid-work.
	turns     int
	closeOnce sync.Once
}

func newMessageStream(
	ctx context.Context,
	cancel context.CancelFunc,
	body io.ReadCloser,
	requestID string,
	maxTotal int64,
	maxEvent int,
) *MessageStream {
	scanner := bufio.NewScanner(body)
	scanner.Buffer(make([]byte, 4096), maxEvent)
	return &MessageStream{
		ctx:       ctx,
		cancel:    cancel,
		body:      body,
		scanner:   scanner,
		requestID: requestID,
		maxTotal:  maxTotal,
		maxEvent:  maxEvent,
	}
}

// RequestID is the upstream x-request-id captured before streaming.
func (stream *MessageStream) RequestID() string {
	if stream == nil {
		return ""
	}
	return stream.requestID
}

// Next returns events in arrival order until the body ends.
//
// A done event is a turn boundary, not the end of the stream: the pinned
// upstream keeps the connection open and starts the next model turn on it
// whenever the previous one called a tool. Hanging up at the first done made
// Berry record a run as finished while the agent was still working. So the
// end of the body is the completion signal: it is io.EOF after at least one
// done, and interrupted before any. Read, parse, and limit failures are
// interrupted (or their own kind) whatever the turn count.
func (stream *MessageStream) Next() (StreamEvent, error) {
	if stream == nil || stream.scanner == nil {
		return StreamEvent{}, errors.New("runtime stream is not configured")
	}
	if stream.eof {
		return StreamEvent{}, stream.finish()
	}
	if err := stream.ctx.Err(); err != nil {
		return StreamEvent{}, stream.contextFailure(err)
	}

	for stream.scanner.Scan() {
		line := stream.scanner.Text()
		stream.total += int64(len(line) + 1)
		stream.frameBytes += len(line) + 1
		if stream.maxTotal < 1 || stream.total > stream.maxTotal ||
			stream.maxEvent < 1 || stream.frameBytes > stream.maxEvent {
			stream.Close()
			return StreamEvent{}, stream.failure(StreamLimit)
		}
		if strings.TrimSuffix(line, "\r") == "" {
			if stream.eventName == "" && len(stream.dataLines) == 0 {
				stream.frameBytes = 0
				continue
			}
			return stream.emit()
		}
		stream.consumeLine(strings.TrimSuffix(line, "\r"))
	}

	if err := stream.scanner.Err(); err != nil {
		stream.Close()
		if contextErr := stream.ctx.Err(); contextErr != nil {
			return StreamEvent{}, stream.contextFailure(contextErr)
		}
		if strings.Contains(err.Error(), "token too long") {
			return StreamEvent{}, stream.failure(StreamLimit)
		}
		return StreamEvent{}, stream.failure(StreamInterrupted)
	}

	stream.eof = true
	if stream.eventName != "" || len(stream.dataLines) > 0 {
		return stream.emit()
	}
	return StreamEvent{}, stream.finish()
}

// emit projects the buffered frame and counts the turn it may close.
func (stream *MessageStream) emit() (StreamEvent, error) {
	event, err := stream.project()
	stream.resetFrame()
	if err != nil {
		stream.Close()
		return StreamEvent{}, err
	}
	if event.Type == EventDone {
		stream.turns++
	}
	return event, nil
}

// finish classifies a body that ended without a read error.
func (stream *MessageStream) finish() error {
	stream.Close()
	if stream.turns > 0 {
		return io.EOF
	}
	return stream.failure(StreamInterrupted)
}

// Close releases the response body and timeout wiring. It is idempotent.
func (stream *MessageStream) Close() error {
	if stream == nil {
		return nil
	}
	var closeErr error
	stream.closeOnce.Do(func() {
		if stream.body != nil {
			closeErr = stream.body.Close()
		}
		if stream.cancel != nil {
			stream.cancel()
		}
	})
	return closeErr
}

func (stream *MessageStream) consumeLine(line string) {
	if line == "" || strings.HasPrefix(line, ":") {
		return
	}
	field, value, found := strings.Cut(line, ":")
	if !found {
		value = ""
	}
	if strings.HasPrefix(value, " ") {
		value = value[1:]
	}
	switch field {
	case "event":
		stream.eventName = value
	case "data":
		stream.dataLines = append(stream.dataLines, value)
	}
}

func (stream *MessageStream) project() (StreamEvent, error) {
	if stream.eventName == "" || len(stream.dataLines) == 0 ||
		len(stream.eventName) > 100 {
		return StreamEvent{}, stream.failure(StreamMalformed)
	}
	data := []byte(strings.Join(stream.dataLines, "\n"))
	if len(data) > stream.maxEvent || !json.Valid(data) {
		return StreamEvent{}, stream.failure(StreamMalformed)
	}

	switch stream.eventName {
	case string(EventChunk):
		var value struct {
			Content string `json:"content"`
		}
		if json.Unmarshal(data, &value) != nil || len(value.Content) > 64*1024 {
			return StreamEvent{}, stream.failure(StreamMalformed)
		}
		return StreamEvent{Type: EventChunk, Content: value.Content}, nil
	case string(EventToolUse):
		var value struct {
			Tool string `json:"tool"`
		}
		if json.Unmarshal(data, &value) != nil || !validEventLabel(value.Tool, 256) {
			return StreamEvent{}, stream.failure(StreamMalformed)
		}
		return StreamEvent{Type: EventToolUse, Tool: value.Tool}, nil
	case string(EventToolResult):
		var value struct {
			Tool  string          `json:"tool"`
			Input json.RawMessage `json:"input"`
		}
		if json.Unmarshal(data, &value) != nil || !validEventLabel(value.Tool, 256) {
			return StreamEvent{}, stream.failure(StreamMalformed)
		}
		if len(value.Input) == 0 {
			value.Input = json.RawMessage("{}")
		}
		if len(value.Input) > 64*1024 || !json.Valid(value.Input) {
			return StreamEvent{}, stream.failure(StreamMalformed)
		}
		return StreamEvent{
			Type:  EventToolResult,
			Tool:  value.Tool,
			Input: append(json.RawMessage(nil), value.Input...),
		}, nil
	case string(EventPhase):
		var value struct {
			Phase  string  `json:"phase"`
			Detail *string `json:"detail"`
		}
		if json.Unmarshal(data, &value) != nil || !validEventLabel(value.Phase, 256) {
			return StreamEvent{}, stream.failure(StreamMalformed)
		}
		if value.Detail != nil && len(*value.Detail) > 4096 {
			return StreamEvent{}, stream.failure(StreamMalformed)
		}
		return StreamEvent{Type: EventPhase, Phase: value.Phase, Detail: value.Detail}, nil
	case string(EventDone):
		var value struct {
			Done  bool  `json:"done"`
			Usage Usage `json:"usage"`
		}
		if json.Unmarshal(data, &value) != nil || !value.Done ||
			value.Usage.InputTokens < 0 || value.Usage.OutputTokens < 0 {
			return StreamEvent{}, stream.failure(StreamMalformed)
		}
		return StreamEvent{Type: EventDone, Usage: value.Usage}, nil
	default:
		if !validEventLabel(stream.eventName, 100) {
			return StreamEvent{}, stream.failure(StreamMalformed)
		}
		return StreamEvent{
			Type:      EventUnknown,
			EventName: stream.eventName,
			Raw:       append(json.RawMessage(nil), data...),
		}, nil
	}
}

func (stream *MessageStream) resetFrame() {
	stream.eventName = ""
	stream.dataLines = stream.dataLines[:0]
	stream.frameBytes = 0
}

func (stream *MessageStream) contextFailure(err error) error {
	if errors.Is(err, context.DeadlineExceeded) {
		return stream.failure(StreamTimeout)
	}
	return stream.failure(StreamInterrupted)
}

func (stream *MessageStream) failure(kind StreamErrorKind) *StreamError {
	return &StreamError{Kind: kind, RequestID: stream.requestID}
}

func validEventLabel(value string, maxBytes int) bool {
	return value != "" && len(value) <= maxBytes && strings.TrimSpace(value) == value
}
